Strict marshalling

jrudolph opened the pull request:
On top of https://github.com/akka/akka-http/pull/3822 gives already about 40 - 50 % perf improvement in grpc_bench.

viktorklang commented on runtime/src/main/scala/akka/grpc/internal/GrpcResponseHelpers.scala:
```diff
@@ -39,6 +39,22 @@ object GrpcResponseHelpers {
+      system: ClassicActorSystemProvider): HttpResponse = {
+    def x = eHandler
+    def f = system
+    if (false) { x; f }
```
what?

jrudolph replied:
The quickest way I found to ignore unused values, will remove it :)

viktorklang commented on runtime/src/main/scala/akka/grpc/internal/GrpcResponseHelpers.scala:
```diff
@@ -39,6 +39,22 @@ object GrpcResponseHelpers {
+    def f = system
+    if (false) { x; f }
+
+    val data = DataFrame(m.serialize(e))
```
Since `data` is an intermediary representation before encodeFrame, I'd probably pass it directly into encodeFrame.

viktorklang commented on runtime/src/main/scala/akka/grpc/internal/GrpcResponseHelpers.scala:
```diff
@@ -39,6 +39,22 @@ object GrpcResponseHelpers {
+
+    val data = DataFrame(m.serialize(e))
+    val strictEntity = HttpEntity.Strict(writer.contentType, writer.encodeFrame(data).data)
+    val trailer = Trailer(GrpcEntityHelpers.trailer(Status.OK).trailers)
```
This could be a cached value, right?

viktorklang commented on runtime/src/main/scala/akka/grpc/internal/GrpcResponseHelpers.scala:
```diff
@@ -39,6 +39,22 @@ object GrpcResponseHelpers {
+    val trailer = Trailer(GrpcEntityHelpers.trailer(Status.OK).trailers)
+    HttpResponse(
+      headers = immutable.Seq(headers.`Message-Encoding`(writer.messageEncoding.name)),
+      entity = strictEntity).addAttribute(AttributeKeys.trailer, trailer)
```
Is it possible to add attributes in the HttpResponse constructor to avoid having to potentially allocate a new HttpResponse immediately?

patriknw commented on runtime/src/main/scala/akka/grpc/internal/GrpcResponseHelpers.scala:
```diff
@@ -39,6 +39,22 @@ object GrpcResponseHelpers {
+    val strictEntity = HttpEntity.Strict(writer.contentType, writer.encodeFrame(data).data)
+    val trailer = Trailer(GrpcEntityHelpers.trailer(Status.OK).trailers)
+    HttpResponse(
+      headers = immutable.Seq(headers.`Message-Encoding`(writer.messageEncoding.name)),
```
headers.`Message-Encoding`(writer.messageEncoding.name) :: Nil

patriknw reviewed (commented):
nice

johanandren reviewed (commented):
👍

jrudolph commented:
Also added the reverse optimization for being able to unmarshal strict entities.

viktorklang commented on runtime/src/main/scala/akka/grpc/javadsl/GrpcMarshalling.scala:
```diff
@@ -55,6 +61,12 @@ object GrpcMarshalling {
+  def unmarshalStream[T](
+      entity: HttpEntity,
+      u: ProtobufSerializer[T],
+      @silent("never used") mat: Materializer,
```
It is used tho?

jrudolph commented:
> Also added the reverse optimization for being able to unmarshal strict entities.

Removed that commit and made a separate change for that in #1372

jrudolph commented:
Trying to fix the test I found that the current HTTP snapshot breaks the Akka HTTP client: if we generate a strict response with a content-length on the server, then we now create a `Default` entity on the client which cannot carry the trailers. I think we should probably disable creating `Default` entities for HTTP/2 for now (the only upside is avoiding a bit of wrapping of Chunks).

jrudolph commented:
This is done now but needs an updated to a version of akka-http that includes https://github.com/akka/akka-http/pull/3841.

jrudolph commented:
Updated to latest released Akka HTTP snapshot. Tests should now succeed.

jrudolph commented on runtime/src/main/scala/akka/grpc/GrpcProtocol.scala:
```diff
@@ -88,6 +88,8 @@ object GrpcProtocol {
       /** Encodes a frame as a part in a chunk stream. */
       encodeFrame: Frame => ChunkStreamPart,
+      /** A shortcut to encode a data frame directly into a ByteString */
+      encodeDataToFrameBytes: ByteString => ByteString,
```
It's a bit unclear for me which parts of the runtime are really internal, and which are used from the generated code and subject to the binary compatibility guarantees.

raboof replied:
yeah, that's tricky - we try to use `@InternalStableApi` for those places we know are used from the generated code, but it's hard to guarantee that is exhaustive. Adding this field looks fine to me, though!

jrudolph commented on runtime/src/main/scala/akka/grpc/internal/GrpcResponseHelpers.scala:
```diff
@@ -27,17 +38,50 @@ import scala.concurrent.{ ExecutionContext, Future }
+      writer: GrpcProtocolWriter,
+      system: ClassicActorSystemProvider): HttpResponse =
+    try {
+      val strictEntity = HttpEntity.Strict(writer.contentType, writer.encodeDataToFrameBytes((m.serialize(e))))
```
This is the main change to create a Strict entity here.

patriknw commented on interop-tests/src/test/scala/akka/grpc/interop/AkkaHttpServerProviderScala.scala:
```diff
@@ -95,17 +96,22 @@ object AkkaHttpServerProviderScala extends AkkaHttpServerProvider with Directive
+            .attribute(AttributeKeys.trailer)
+            .map(_.headers)
+            .getOrElse(Vector.empty)
+            .map(e => RawHeader(e._1, e._2))
```
could these map+get+map be collapsed to one? maybe not in a hot path?

jrudolph replied:
That's test code right now.

patriknw reviewed (approved):
LGTM
